File: src/slack_claude_bridge/job.py
"""Outbound entry points: start a Claude job locally and push its result into
Slack so you can continue it from your phone.

- ``slack-claude-job "<prompt>"``  — run a prompt headless via the same
  ClaudeRunner the bridge uses, then post the result to Slack and seed the
  thread with the new session id.
- ``slack-claude-notify --session <id>`` — push arbitrary text for an existing
  session (for a Claude Code Stop hook, or any other integration).

Both create a Slack thread mapped to a Claude session, so replying in that
thread (mention the bot) resumes it via --resume.
"""

from __future__ import annotations

import argparse
import logging
import sys

from .claude_runner import ClaudeRunner
from .config import Config, ConfigError
from .notify import push_result
from .store import SessionStore

logger = logging.getLogger("slack_claude_bridge.job")


def _resolve_channel(config: Config, override: str | None) -> str:
    channel = override or config.notify_channel
    if not channel:
        raise SystemExit(
            "No target channel. Pass --channel C0123 or set SLACK_NOTIFY_CHANNEL in .env."
        )
    return channel


def run_job_main() -> None:
    """`slack-claude-job` — run a prompt headless, then push the result."""
    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s: %(message)s")

    parser = argparse.ArgumentParser(
        prog="slack-claude-job",
        description="Run a Claude Code job locally and post the result to Slack.",
    )
    parser.add_argument("prompt", nargs="+", help="The task for Claude (quote it).")
    parser.add_argument("--channel", help="Slack channel ID to post to (overrides SLACK_NOTIFY_CHANNEL).")
    parser.add_argument("--workdir", help="Directory Claude runs in (overrides CLAUDE_WORKDIR).")
    parser.add_argument("--title", help="Custom header for the Slack message.")
    args = parser.parse_args()

    try:
        config = Config.load()
    except ConfigError as exc:
        raise SystemExit(f"Configuration error: {exc}")

    channel = _resolve_channel(config, args.channel)
    prompt = " ".join(args.prompt)

    runner = ClaudeRunner(
        binary=config.claude_bin,
        workdir=args.workdir or config.workdir,
        permission_mode=config.permission_mode,
        model=config.model,
        extra_args=config.extra_args,
        timeout=config.timeout,
    )

    logger.info("Running Claude job (this may take a while)…")
    result = runner.run_new(prompt)

    store = SessionStore(config.db_path)
    thread_ts = push_result(
        config=config,
        store=store,
        channel=channel,
        session_id=result.session_id,
        text=result.text,
        is_error=result.is_error,
        title=args.title,
    )
    print(f"Posted to {channel} (thread_ts={thread_ts}, session={result.session_id}).")
    if result.is_error:
        raise SystemExit(1)


def notify_main() -> None:
    """`slack-claude-notify` — push text for an existing session id."""
    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s: %(message)s")

    parser = argparse.ArgumentParser(
        prog="slack-claude-notify",
        description="Post text to Slack for an existing Claude session and seed the thread.",
    )
    parser.add_argument("--session", required=True, help="Claude session id to resume on reply.")
    parser.add_argument("--channel", help="Slack channel ID (overrides SLACK_NOTIFY_CHANNEL).")
    parser.add_argument("--title", help="Custom header for the Slack message.")
    parser.add_argument("--error", action="store_true", help="Mark the result as an error.")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("--text", help="Result text. If omitted, read from stdin.")
    group.add_argument("--file", help="Read result text from this file.")
    args = parser.parse_args()

    try:
        config = Config.load()
    except ConfigError as exc:
        raise SystemExit(f"Configuration error: {exc}")

    channel = _resolve_channel(config, args.channel)

    if args.text is not None:
        text = args.text
    elif args.file:
        with open(args.file, encoding="utf-8") as fh:
            text = fh.read()
    else:
        text = sys.stdin.read()

    store = SessionStore(config.db_path)
    thread_ts = push_result(
        config=config,
        store=store,
        channel=channel,
        session_id=args.session,
        text=text,
        is_error=args.error,
        title=args.title,
    )
    print(f"Posted to {channel} (thread_ts={thread_ts}, session={args.session}).")

File: src/bin/job.rs
//! `slack-claude-job` — run a Claude job locally, then push its result to Slack
//! and seed the thread (mirror of `job.py:run_job_main`).
//!
//! Usage: slack-claude-job [--channel C0123] [--workdir DIR] [--title T] <prompt...>

use std::process::exit;

use tokio::runtime::Runtime;

use slack_claude_bridge::claude_runner::ClaudeRunner;
use slack_claude_bridge::config::Config;
use slack_claude_bridge::notify::push_result;
use slack_claude_bridge::slack::{build_client, RealSlack};
use slack_claude_bridge::store::SessionStore;

fn main() {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::try_from_default_env().unwrap_or_else(|_| "info".into()),
        )
        .init();

    let mut channel: Option<String> = None;
    let mut workdir: Option<String> = None;
    let mut title: Option<String> = None;
    let mut prompt_parts: Vec<String> = Vec::new();

    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--channel" => channel = args.next(),
            "--workdir" => workdir = args.next(),
            "--title" => title = args.next(),
            _ => prompt_parts.push(arg),
        }
    }

    if prompt_parts.is_empty() {
        eprintln!("usage: slack-claude-job [--channel C] [--workdir DIR] [--title T] <prompt...>");
        exit(2);
    }
    let prompt = prompt_parts.join(" ");

    let config = Config::load().unwrap_or_else(|e| {
        eprintln!("Configuration error: {e}");
        exit(2);
    });
    let channel = channel
        .or_else(|| config.notify_channel.clone())
        .unwrap_or_else(|| {
            eprintln!(
                "No target channel. Pass --channel C0123 or set SLACK_NOTIFY_CHANNEL in .env."
            );
            exit(2);
        });

    let runner = ClaudeRunner {
        binary: config.claude_bin.clone(),
        workdir: workdir.unwrap_or_else(|| config.workdir.clone()),
        permission_mode: config.permission_mode.clone(),
        model: config.model.clone(),
        extra_args: config.extra_args.clone(),
        timeout: config.timeout,
    };

    eprintln!("Running Claude job (this may take a while)…");
    let result = runner.run_new(&prompt);

    let rt = Runtime::new().expect("tokio runtime");
    let (client, token) = build_client(&config.bot_token).unwrap_or_else(|e| {
        eprintln!("Slack client error: {e}");
        exit(2);
    });
    let slack = RealSlack::new(client, token, rt.handle().clone());
    let store = SessionStore::new(&config.db_path);

    let thread_ts = push_result(
        &slack,
        &store,
        &channel,
        result.session_id.as_deref(),
        &result.text,
        result.is_error,
        title.as_deref(),
    );
    println!(
        "Posted to {channel} (thread_ts={thread_ts}, session={}).",
        result.session_id.as_deref().unwrap_or("?")
    );
    if result.is_error {
        exit(1);
    }
}
